openingd/dualopend.c

```diff
@@ -95,8 +123,1303 @@ struct state {
+				   " max_htlc_value_in_flight_msat is %s,"
+				   " channel capacity is %s, which is below %s",
+				   type_to_string(tmpctx, struct amount_sat, &amt),
+				   type_to_string(tmpctx, struct amount_sat,
+						  &remoteconf->channel_reserve),
+				   type_to_string(tmpctx, struct amount_sat,
+						  &state->localconf.channel_reserve),
+				   type_to_string(tmpctx, struct amount_msat,
+						  &remoteconf->max_htlc_value_in_flight),
+				   type_to_string(tmpctx, struct amount_sat,
+						  &capacity),
+				   type_to_string(tmpctx, struct amount_msat,
+						  &state->min_effective_htlc_capacity));
+		return false;
+	}
+
+	/* We don't worry about how many HTLCs they accept, as long as > 0! */
+	if (remoteconf->max_accepted_htlcs == 0) {
+		negotiation_failed(state, am_opener,
+				   "max_accepted_htlcs %u invalid",
+				   remoteconf->max_accepted_htlcs);
+		return false;
+	}
+
+	/* BOLT #2:
+	 *
+	 * The receiving node MUST fail the channel if:
+	 *...
+	 *  - `max_accepted_htlcs` is greater than 483.
+	 */
+	if (remoteconf->max_accepted_htlcs > 483) {
+		negotiation_failed(state, am_opener,
+				   "max_accepted_htlcs %u too large",
+				   remoteconf->max_accepted_htlcs);
+		return false;
+	}
+
+	/* BOLT #2:
+	 *
+	 * The receiving node MUST fail the channel if:
+	 *...
+	 *  - `dust_limit_satoshis` is greater than `channel_reserve_satoshis`.
+	 */
+	if (amount_sat_greater(remoteconf->dust_limit,
+			       remoteconf->channel_reserve)) {
+		negotiation_failed(state, am_opener,
+				   "dust_limit_satoshis %s"
+				   " too large for channel_reserve_satoshis %s",
+				   type_to_string(tmpctx, struct amount_sat,
+						  &remoteconf->dust_limit),
+				   type_to_string(tmpctx, struct amount_sat,
+						  &remoteconf->channel_reserve));
+		return false;
+	}
+
+	return true;
+}
+
+static bool is_segwit_output(struct wally_tx_output *output,
+			     const u8 *redeemscript)
+{
+	u8 *wit_prog;
+	if (tal_bytelen(redeemscript) > 0)
+		wit_prog = cast_const(u8 *, redeemscript);
+	else
+		wit_prog = cast_const(u8 *, wally_tx_output_get_script(tmpctx, output));
+
+	return is_p2wsh(wit_prog, NULL) || is_p2wpkh(wit_prog, NULL);
+}
+
+static struct wally_psbt *
+fetch_next_moves(struct state *state, const struct wally_psbt *psbt)
+{
+	u8 *msg;
+	u16 msg_type;
+	struct wally_psbt *updated_psbt;
+
+	/* Go ask lightningd what other changes we've got */
+	msg = towire_dual_open_psbt_changed(NULL, &state->channel_id, psbt);
+
+	wire_sync_write(REQ_FD, take(msg));
+	msg = wire_sync_read(tmpctx, REQ_FD);
+
+	msg_type = fromwire_peektype(msg);
+	if (msg_type == WIRE_DUAL_OPEN_FAIL) {
+		char *err;
+		if (!fromwire_dual_open_fail(msg, msg, &err))
+			master_badmsg(msg_type, msg);
+
+		peer_failed(state->pps, &state->channel_id, "%s", err);
+	} else if (msg_type == WIRE_DUAL_OPEN_PSBT_CHANGED) {
+		struct channel_id unused;
+		if (!fromwire_dual_open_psbt_changed(state, msg, &unused, &updated_psbt)) {
+			master_badmsg(msg_type, msg);
+			return NULL;
```

Comment:
master_badmsg does not return?

Reply:
it exits, which is a super-return. will update.